ci: add arm64 runner

clason opened the pull request:
Problem: No tests on Linux Arm.

Solution: Add tests on Linux Arm with `RelWithDebInfo`.   Remove duplicate and low-leverage `build-types` job.

clason committed:
ci(tests): add arm64 runner

Problem: Linux `aarch64`/`arm64` builds are not tested.

Solution: Add `ubuntu-arm` runners to test matrix (using
`RelWithDebInfo` build).

clason commented:
Looks good, only issue is permissions with the `vim.ui.open()` tests around nonexistent files (permission issues, I think we had similar on Cirrus). At least the first (validation) test sounds like an actually relevant issue and should likely be fixed before providing releases?

Follow-up question would be whether and how we want to test this platform (add to matrix? switch one of the other (`release`) tests to ARM?)

dundargoc commented on runtime/doc/support.txt:
```diff
@@ -13,6 +13,7 @@ Supported platforms					 *supported-platforms*
 
 `System`          `Tier`      `Versions`                  `Tested versions`
 Linux            1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
```
I guess add x86-64 or arm64 or whatever the kids use these day?

dundargoc commented on .github/workflows/test.yml:
```diff
@@ -107,6 +107,7 @@ jobs:
             { runner: ubuntu-24.04, os: ubuntu, flavor: asan, cc: clang, flags: -D ENABLE_ASAN_UBSAN=ON },
             { runner: ubuntu-24.04, os: ubuntu, flavor: tsan, cc: clang, flags: -D ENABLE_TSAN=ON },
             { runner: ubuntu-24.04, os: ubuntu, flavor: release, cc: gcc, flags: -D CMAKE_BUILD_TYPE=Release },
+            { runner: ubuntu-24.04-arm, os: ubuntu, flavor: arm, cc: gcc, flags: -D CMAKE_BUILD_TYPE=RelWithDebInfo },
```
You can remove the `RelWithDebInfo` build in the `build-types` job. In fact, with this I'd say we can remove the `build-types` job entirely as MinSizeRel or Ninja Multi-Config isn't really important enough to warrant a CI job.

dundargoc replied:
Either do it in this PR or I can do it in a follow up.

clason replied:
@bfredl You'd need to remove the `build-types` job from the list of required checks then.

clason committed:
ci(tests): remove build-types jobs

Problem: Some CI jobs are redundant: `RelWithDebInfo` is already tested
on Linux-Arm64; `MinSizeRel` and Ninja Multi Config are not sufficiently
relevant in practice to spend CI cycles on.

Solution: Remove `build-types` job.

clason commented:
Let's get the tests in now; release workflow will be a follow-up in the next days.

zeertzjq commented:
Why remove build-types jobs?

clason commented:
Problem: Some CI jobs are redundant: `ResWithDebInfo` is already tested
on Linux-Arm64; `MinSizeRel` and Ninja Multi Config are not sufficiently
relevant in practice to spend CI cycles on.

Solution: Remove `build-types` job.

zeertzjq commented on runtime/doc/support.txt:
```diff
@@ -11,15 +11,16 @@ Support                                                              *support*
-MinGW            2      MinGW-w64
-Windows 64-bit   3      < Windows 10 Version 1809
+`System`            `Tier`      `Versions`                  `Tested versions`
+Linux (x864_64)    1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
```
```suggestion
Linux (x86_64)     1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
```

zeertzjq commented on runtime/doc/support.txt:
```diff
@@ -11,15 +11,16 @@ Support                                                              *support*
-MinGW            2      MinGW-w64
-Windows 64-bit   3      < Windows 10 Version 1809
+`System`            `Tier`      `Versions`                  `Tested versions`
+Linux (x864_64)    1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
+Linux (aarch64)    1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
```
```suggestion
Linux (x86_64)     1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
Linux (arm64)      1      >= 2.6.32, glibc >= 2.12     Ubuntu 24.04
```

clason replied:
That arch was intentional. Can you tell my _why_ we should use one over the other?

zeertzjq replied:
No

clason replied:
wat

zeertzjq replied:
Well I'm not sure either. But GitHub calls it arm64.

clason replied:
So does Debian, so let's go with that. (It's also shorter.)

adamency commented:
# PSA

This PR adds ARM64 builds to all future releases (i.e. after this PR was merged).

zeertzjq commented:
No, it's #32194
